### ISSUE TITLE
[new rule] Disallow usage of unnecessary expectations / matchers

### ISSUE THREAD

**Issue by maks-rafalko** [jest-community/eslint-plugin-jest#1616]
Currently, the following code doesn't produce any errors neither from TS, nor from `@typescript/eslint` and `eslint-plugin-jest`:

```ts
const getNames = (): string[] => {
  return ['one', 'two'];
};

test('...', () => {
  const names = getNames();

  expect(names).toBeDefined();
});
```

Note the `string[]` return type of `getNames()` function. It's literally always "defined" so `.toBeDefined()` does not make any sense here, but neither TS nor ESLint complain about it, giving developers a false feeling of good tests.

Actual result: no errors, tests obviously pass while they make no sense.

Expected result: linter error saying `Variable "names" has "string[]" type, checking it with "toBeDefined()" is useless and laways pass. Consider using another matcher"`

Ddo you think `eslint-plugin-jest` is the right place to have a new rule, checking already known types of variables and checking matchers, disallowing useless usage?

Some expiration for implementation can be found in `typescript-eslint`'s [`no-unnecessary-type-assertion`](https://github.com/typescript-eslint/typescript-eslint/blob/8f08ee54d1389666b3f582853a8b6809d34d70f8/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts)

**Comment by G-Rath** [comment]
@maks-rafalko sorry for the late reply - I'd be willing to explore this especially since you've already got working rules. Would you be open to porting them over to here?

**Comment by G-Rath** [comment]
I've opened #1882 with a type-based rule for this, which should be good to go I just need to finish off some docs and tests - if anyone's watching this issue, it'd be great if folks could give the rule a go and let me know if it works. I've not got a lot experience with type-based rules and don't have any test suites that are actually using the related matchers 😅 

This is the script I often use to test changes:

```bash
#!/usr/bin/env bash
p="/path/to/codebase/node_modules/eslint-plugin-jest"

yarn run build

rm -rf "$p/lib"
cp -r "./lib/" "$p"
```
